### PR TITLE
Returns errors from DB methods, instead of `found` bools and critical log events

### PR DESCRIPTION
**Pull request by joeldudley**
### Why is this change needed?

- Provide a description and a link to the underlying ticket

### What changes were made as part of this PR:

- Provide a high level list of the changes made
- List key areas for the reviewer 

### :rotating_light: Definition of done :rotating_light:
- [ ] Unit tests added to cover new or changed functionality 
- [ ] Docs pages updated to cover new or changed functionality
- [ ] [Changelog.md](https://github.com/obscuronet/go-obscuro/blob/main/docs/testnet/changelog.md) updated

### CODE REVIEW

File: go/host/db/batches.go [hunk 1]
```diff
@@ -2,106 +2,121 @@ package db
...
 	}
 	data, err := db.kvStore.Get(batchHeaderKey(hash))
 	if err != nil {
-		db.logger.Crit("could not retrieve batch header.", log.ErrKey, err)
+		return nil, err
 	}
 	if len(data) == 0 {
-		return nil, false
+		return nil, errutil.ErrNotFound
 	}
```
Comment (otherview) [inline go/host/db/batches.go:76]: Why 2 operations ? Has and then Get ?

File: go/host/db/batches.go [hunk 2]
```diff
@@ -2,106 +2,121 @@ package db
...
 	if !f {
-		return nil, false
+		return nil, errutil.ErrNotFound
 	}
 	value, err := db.kvStore.Get(headBatch)
 	if err != nil {
-		db.logger.Crit("could not retrieve head batch.", log.ErrKey, err)
+		return nil, err
 	}
```
Comment (otherview) [inline go/host/db/batches.go:96]: Similar has and get operations. Should this be just one ?

File: go/host/rpc/clientapi/client_api_obscuroscan.go [hunk 3]
```diff
@@ -26,21 +29,24 @@ func NewObscuroScanAPI(host host.Host) *ObscuroScanAPI {
...
 func (api *ObscuroScanAPI) GetBlockHeaderByHash(blockHash gethcommon.Hash) (*types.Header, error) {
-	blockHeader, found := api.host.DB().GetBlockHeader(blockHash)
-	if !found {
+	blockHeader, err := api.host.DB().GetBlockHeader(blockHash)
+	if errors.Is(err, errutil.ErrNotFound) {
 		return nil, fmt.Errorf("no block with hash %s is stored", blockHash)
 	}
+	if err != nil {
+		return nil, err
+	}
```
Comment (otherview) [inline go/host/rpc/clientapi/client_api_obscuroscan.go:38]:
```suggestion
if err != nil {
if errors.Is(err, errutil.ErrNotFound) {
		return nil, fmt.Errorf("no block with hash %s is stored", blockHash)
	}
		return nil, err
	}
	
	
```

File: go/host/rpc/clientapi/client_api_obscurotest.go [hunk 4]
```diff
@@ -19,15 +22,17 @@ func NewTestAPI(host host.Host) *TestAPI {
...
-	if !found {
-		return 0
+func (api *TestAPI) BlockNumber() (hexutil.Uint64, error) {
+	head, err := api.host.DB().GetHeadBlockHeader()
+	if errors.Is(err, errutil.ErrNotFound) {
+		return 0, nil
+	}
+	if err != nil {
+		return 0, err
 	}
```
Comment (otherview) [inline go/host/rpc/clientapi/client_api_obscurotest.go:32]: No need for a different if condition

File: go/host/db/batches.go [hunk 5]
```diff
@@ -2,106 +2,121 @@ package db
...
 }
 
 // AddBatchHeader adds a batch's header to the known headers
-func (db *DB) AddBatchHeader(header *common.Header, txHashes []common.TxHash) {
+func (db *DB) AddBatchHeader(header *common.Header, txHashes []common.TxHash) error {
 	b := db.kvStore.NewBatch()
-	db.writeBatchHeader(header)
+	err := db.writeBatchHeader(header)
+	if err != nil {
+		return err
```
Comment (otherview) [inline go/host/db/batches.go:30]: Matter of preference, but I typically wrap the errors with a reason in methods that may return more than one error.

File: go/host/db/batches.go (hunk 1 above)
Comment (joeldudley) [reply go/host/db/batches.go:76]: Yeah, that's the way it works apparently. Tried removing the `Has`, didn't work.

Comment (otherview) [reply go/host/db/batches.go:76]: I mean why use the two operations when `readBatchHeader` returns the value?
In other words, could the Has calling be skipped and just do the Get call ?

Comment (joeldudley) [reply go/host/db/batches.go:76]: But this is inside `readBatchHeader`?

Comment (otherview) [reply go/host/db/batches.go:76]: Why is the `readBatchHeader` method (yes, readBatchHeader is the method that we're discussing) doing Has (`db.kvStore.Has(batchHeaderKey(hash))`)  and then doing a Get (`db.kvStore.Get(batchHeaderKey(hash))`)? 
Can it not just do the Get ? 

Give me a shout if I'm not explaining myself properly :)

Comment (joeldudley) [reply go/host/db/batches.go:76]: It's not clear to me either (just copy and pasting stuff to a large extent :D ), but try removing it and you'll see. I assume Matt went through this when he originally wrote it.

Comment (joeldudley) [reply go/host/db/batches.go:76]: Happy for us to investigate further. But feel this is really nothing to do with my PR. It was like that before.

Comment (otherview) [reply go/host/db/batches.go:76]: It's a show-stopper for me to be honest. But I understand that it sidetracks this feature. TODO + create issue ?

Comment (joeldudley) [reply go/host/db/batches.go:76]: Yeah, TODO + issue is fine. I'll add.

I understand your concerns, but I think a PR review should be broadly limited to what's being changed, not whether your code changes something near something crap :)

File: go/host/db/batches.go (hunk 5 above)
Comment (joeldudley) [reply go/host/db/batches.go]: Done

File: go/host/rpc/clientapi/client_api_eth.go [hunk 6]
```diff
@@ -34,8 +36,9 @@ func (api *EthereumAPI) ChainId() (*hexutil.Big, error) { //nolint:stylecheck,re
 // BlockNumber returns the height of the current head rollup.
 // # TODO - #718 - Switch to returning height based on current batch.
 func (api *EthereumAPI) BlockNumber() hexutil.Uint64 {
-	header, found := api.host.DB().GetHeadRollupHeader()
-	if !found {
+	header, err := api.host.DB().GetHeadRollupHeader()
+	if err != nil {
+		// This error may be nefarious, but unfortunately the Eth API doesn't allow us to return an error.
```
Comment (tudor-malene) [inline go/host/rpc/clientapi/client_api_eth.go:41]: we should log the error .

**Review by tudor-malene — APPROVED** [review]
lgtm
(one minor thing I spotted)

**Pull request merged**